README.md
# ble2mqtt
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](LICENSE)
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/sebbo2002/ble2mqtt/Release?style=flat-square)](https://github.com/sebbo2002/ble2mqtt/actions)

## 🧐 What's this?

ble2mqtt is a small script that **writes data from Bluetooth-LE devices to MQTT Topics**. Unlike many other
implementations it **only listens to the broadcasts** of the devices, there is **no active polling**. This has the
advantage that it is more stable, the sensors consumes less energy and mostly the range is higher. The disadvantage is
that, depending on the device, individual values may be missing if they are not transmitted by broadcast. If these
values are required, they would still have to be retrieved by polling. However, the most important data is usually
transmitted via broadcasts, so that less important data (e.g. the battery status of the device) can also be requested
less frequently.

## 💾 Installation
This module is based on noble. Therefore the same requirements are necessary here as well.

- **for macOS**: install [Xcode](https://itunes.apple.com/ca/app/xcode/id497799835?mt=12)
- **for Ubuntu/Debian/Raspbian**: run `sudo apt-get install bluetooth bluez libbluetooth-dev libudev-dev`
- **for Fedora / Other-RPM based**: run `sudo yum install bluez bluez-libs bluez-libs-devel`
- for **Windows and other systems** have a look at [the noble project](https://github.com/noble/noble/blob/master/README.md#prerequisites), they may already have some instructions for you


After you have met the requirements for noble, you can install bte2mqtt:

```bash
npm install -g @sebbo2002/ble2mqtt
```

If you want to run ble2mqtt without root privileges, you have to give the node binary the right cap_net_raw under Linux,
otherwise no Bluetooth advertisements can be received. How to do this is described [here](https://github.com/abandonware/noble#running-without-rootsudo-linux-specific) quite well.

## 🔧 Configuration
| Environment Variable | Default Value      | Description                                                                               |
|:-------------------- |:------------------ |:----------------------------------------------------------------------------------------- |
| `BROKER_URL`         | `mqtt://localhost` | MQTT-Broker URI                                                                           |
| `TOPIC_PREFIX`       | `ble2mqtt`         | Prefix used to generate MQTT topics                                                       |
| `CACHE_TTL`          | `5000`             | Cache TTL in ms, set to `0` to disable cache and publish every single message it gets     |
| `RETAIN_FLAG`        | `0`                | Set the retain flag for MQTT messages? Default is no, set to `1` to enable.               |
| `WHITELIST`          | -                  | Comma seperated list of device UUIDs or addresses. If set, all other devices are ignored. |
| `MONITORING_ID`      | -                  | Added to monitoring topics like `pid` or `uptime` to distinguish multiple agents          |
| `WATCHDOG_SUICIDE`   | `0`                | Setting this to `1` allows the internal Watchdog to kill ble2mqtt if something goes wrong |

## ⌨️ Usage

If you have ble2mqtt installed, you can start the tool. The best thing to do is to set `--debug` first, so you can see
what happens. When the script reports `Connections established` it will start. If not make sure that no other programs
access your Bluetooth adapter. This can cause problems.

![ble2mqtt screenshot](https://d.sebbo.net/ble2mqtt-91RDXKYyzGHryca265JdRtznbovUo7d9ibyVxNwPfy5b1FMkmX4wNhrehMMOpEdlE1GHPVXvkMTwPqmYKGoTSkeiopmVuy2G3Dz3.png)

ble2mqtt now collects data from devices that send it out via Bluetooth LE. This data is parsed and forwarded in MQTT
topics. Besides the general attributes there are also devices that reveal more about themselves. These are mostly
sensors. If they are implemented in ble2mqtt, these values are forwarded as well. A list with all supported devices can
be found below.

#### General attributes
Event if there's no implementation for your device you can use this module to get some basic information. You can still
use other implementations to poll values in parallel. These general attributes are:
- `name`
- `address`
- `uuid`
- `lastSeen`
- `rssi`

#### Monitoring
ble2mqtt supports several topics that can be used to monitor the process. They are updated every times the script starts.
- `pid`
- `version`
- `uptime`
- `watchdogTimeout`

#### Internal Cache
To avoid flooding the MQTT-Broker with a huge amount of messages, there is a limit of one update per topic per 5
seconds. The limit can be changed or deactivated with the environment variable `CACHE_TTL`.


## 📡 Supported Devices

Here's a list of devices which are supported. Other devices as listed here will most likely only support the basic
attributes.

<table>
    <tr>
        <td><img src="https://d.sebbo.net/HHCCJCY01-sltHKnA3SmaAoB2B1ZTv2NwT9YPcDjno7tVF0PbOzjyb6lCb0IWGEjQVD4xAAG5ooqVito3tktrqyoOEwK7BTVkqEnZFHYrRpgaQ.jpeg" width="100" alt="LYWSDCGQ"></td>
        <td><b>Mi Flora Sensor</b><br />HHCCJCY01</td>
    </tr>
    <tr>
        <td><b>Values:</b> (without basics)</td>
        <td>temperature, moisture, illuminance, fertility</td>
    </tr>
    <tr>
        <td><b>Update interval:</b></td>
        <td>~ 1x / min</td>
    </tr>
</table>

<table>
    <tr>
        <td><img src="https://d.sebbo.net/LYWSDCGQ-jIYfxXMRFf0jm3XmPaONLG0RWofIr2JL9JGpshkGmGvMhzW3yMXE8gVI7Xut2Osz4oUkJnES2iB38IVr74HP1kSZRnqsu2TC7RxT.jpeg" width="100" alt="LYWSDCGQ"></td>
        <td><b>Xiaomi Thermometer / Hygrometer</b><br />LYWSDCGQ</td>
    </tr>
    <tr>
        <td><b>Values:</b> (without basics)</td>
        <td>temperature, humidity, battery</td>
    </tr>
    <tr>
        <td><b>Update interval:</b></td>
        <td>~ 20x / min</td>
    </tr>
</table>

<table>
    <tr>
        <td><img src="https://d.sebbo.net/Shared-Image-2020-11-22-18-55-51-w01pqMCZdUPOQreGg1k446bwArsD4Fvw8CmlTy6kSqewRC2V5htTkwO9Rj04Vx26v2UVDMAG179bZhCcajqjNt4E6f4VPplyCLPz.png" width="100" alt="LYWSDCGQ"></td>
        <td><b>SwitchBot Bot</b><br />WoHand</td>
    </tr>
    <tr>
        <td><b>Values:</b> (without basics)</td>
        <td>switchAddOnUsed, state, battery</td>
    </tr>
    <tr>
        <td><b>Update interval:</b></td>
        <td><i>?</i></td>
    </tr>
</table>

<table>
    <tr>
        <td><img src="https://d.sebbo.net/Shared-Image-2020-11-22-18-56-39-pOkUJpmde1r4lRSsOl02QqX6RcswTBf4FY4uGaDwTjyTOa0CArzZJ1YVdQniHJMOE9GLBcL8fSQyFwjmf0KfsxZvWzmgSMP2vREv.png" width="100" alt="LYWSDCGQ"></td>
        <td><b>SwitchBot Meter</b><br />WoSensorTH</td>
    </tr>
    <tr>
        <td><b>Values:</b> (without basics)</td>
        <td>temperature, humidity, battery</td>
    </tr>
    <tr>
        <td><b>Update interval:</b></td>
        <td><i>?</i></td>
    </tr>
</table>

<table>
    <tr>
        <td><img src="https://d.sebbo.net/Shared-Image-2020-11-22-18-57-04-ypZ9SYzljJl8rGj8NVPuadVQLwdAsn6fLsliFbkN8BNdGtwRBQysTv64ywbTIg07UD2BwppJI9wwQqTx1nWyHL6IcLk0dK9SMQm8.png" width="100" alt="LYWSDCGQ"></td>
        <td><b>SwitchBot Curtain</b><br />WoCurtain</td>
    </tr>
    <tr>
        <td><b>Values:</b> (without basics)</td>
        <td>calibrated, battery, position, lightLevel</td>
    </tr>
    <tr>
        <td><b>Update interval:</b></td>
        <td><i>?</i></td>
    </tr>
</table>


## 🤷 Frequently asked questions
### How do I determine which Bluetooth adapter the script should use?
- `hcitool dev`
- `NOBLE_HCI_DEVICE_ID=1` (hci1 used)

### Something doesn't work. How can I get more information about what's going on inside the script?
- `ble2mqtt --debug`

### After some time ble2mqtt does not work anymore, what can be the reason?
The problem is known, but unfortunately I do not know yet what the exact reason is. On some hosts I have the problem,
on others not. A restart will reliably solve the problem. If ble2mqtt was started in such a way that it is automatically
restarted when it terminates (e.g. via `forever`), `WATCHDOG_SUICIDE` can be used to allow the watchdog to terminate
ble2mqtt when no more data is received. This works very reliably for me.

### I have a large area to cover. How do I scale ble2mqtt with multiple agents?
You can run the script on multiple hosts and run it in parallel. If you set the same topic, it makes no difference which
instance received the beacons from the device. Set `MONITORING_ID` if you want to distinguish the monitoring topics.


## 📚 Credits
- [noble](https://github.com/abandonware/noble#readme)
- [mqtt.js](https://github.com/mqttjs/MQTT.js)

## 👩‍⚖️ Copyright & License
Copyright (c) Sebastian Pekarek under the [MIT license](LICENSE).


CHANGELOG.md
## [3.1.3-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.2...v3.1.3-develop.1) (2023-01-01)

## [3.1.2](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.1...v3.1.2) (2022-10-10)

## [3.1.2-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.1...v3.1.2-develop.1) (2022-10-10)

## [3.1.1](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.0...v3.1.1) (2022-09-15)

## [3.1.1-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.0...v3.1.1-develop.1) (2022-09-14)

# [3.1.0](https://github.com/sebbo2002/ble2mqtt/compare/v3.0.0...v3.1.0) (2022-08-26)


### Bug Fixes

* **Switchbot:** Add new service UUID ([ad6c0bf](https://github.com/sebbo2002/ble2mqtt/commit/ad6c0bfc773a566feff1b62ef83ea022628a64bd))


### Features

* **Switchbot:** Add `moving` flag ([3bd5f20](https://github.com/sebbo2002/ble2mqtt/commit/3bd5f2008592cd1a15cb79ecef3b6253ae2c4973))

# [3.1.0-develop.3](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.0-develop.2...v3.1.0-develop.3) (2022-08-24)

# [3.1.0-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v3.1.0-develop.1...v3.1.0-develop.2) (2022-08-16)

# [3.1.0-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v3.0.0...v3.1.0-develop.1) (2022-07-29)


### Bug Fixes

* **Switchbot:** Add new service UUID ([ad6c0bf](https://github.com/sebbo2002/ble2mqtt/commit/ad6c0bfc773a566feff1b62ef83ea022628a64bd))


### Features

* **Switchbot:** Add `moving` flag ([3bd5f20](https://github.com/sebbo2002/ble2mqtt/commit/3bd5f2008592cd1a15cb79ecef3b6253ae2c4973))

# [3.0.0](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.8...v3.0.0) (2022-07-25)


### Build System

* Native ESM support ([7b86a4f](https://github.com/sebbo2002/ble2mqtt/commit/7b86a4f1187c387a3a5792e1fb72d822b04e3631))


### BREAKING CHANGES

* Only Support for node.js ^12.20.0 || >=14.13.1

# [3.0.0-develop.4](https://github.com/sebbo2002/ble2mqtt/compare/v3.0.0-develop.3...v3.0.0-develop.4) (2022-07-25)

# [3.0.0-develop.3](https://github.com/sebbo2002/ble2mqtt/compare/v3.0.0-develop.2...v3.0.0-develop.3) (2022-07-22)

# [3.0.0-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v3.0.0-develop.1...v3.0.0-develop.2) (2022-07-16)

# [3.0.0-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.9-develop.2...v3.0.0-develop.1) (2022-07-16)


### Build System

* Native ESM support ([7b86a4f](https://github.com/sebbo2002/ble2mqtt/commit/7b86a4f1187c387a3a5792e1fb72d822b04e3631))


### BREAKING CHANGES

* Only Support for node.js ^12.20.0 || >=14.13.1

## [2.0.9-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.9-develop.1...v2.0.9-develop.2) (2022-07-01)

## [2.0.9-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.8...v2.0.9-develop.1) (2022-06-22)

## [2.0.8](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.7...v2.0.8) (2022-06-21)

## [2.0.8-develop.4](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.8-develop.3...v2.0.8-develop.4) (2022-06-13)

## [2.0.8-develop.3](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.8-develop.2...v2.0.8-develop.3) (2022-06-10)

## [2.0.8-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.8-develop.1...v2.0.8-develop.2) (2022-06-08)

## [2.0.8-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.7...v2.0.8-develop.1) (2022-06-07)

## [2.0.7](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.6...v2.0.7) (2022-06-01)


### Reverts

* Revert "ci: Remove docker setup" ([14a5226](https://github.com/sebbo2002/ble2mqtt/commit/14a5226c828f173fe23b1d0033f825c73e848038))
* Revert "ci: Remove docker setup" ([655068b](https://github.com/sebbo2002/ble2mqtt/commit/655068b3b9c6139181ae87421db5f8144fae3e18))

## [2.0.7-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.7-develop.1...v2.0.7-develop.2) (2022-05-20)


### Reverts

* Revert "ci: Remove docker setup" ([655068b](https://github.com/sebbo2002/ble2mqtt/commit/655068b3b9c6139181ae87421db5f8144fae3e18))

## [2.0.7-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.6...v2.0.7-develop.1) (2022-05-20)


### Reverts

* Revert "ci: Remove docker setup" ([14a5226](https://github.com/sebbo2002/ble2mqtt/commit/14a5226c828f173fe23b1d0033f825c73e848038))

## [2.0.6](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.5...v2.0.6) (2022-05-14)

## [2.0.6-develop.4](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.6-develop.3...v2.0.6-develop.4) (2022-05-14)

## [2.0.6-develop.3](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.6-develop.2...v2.0.6-develop.3) (2022-05-03)

## [2.0.6-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.6-develop.1...v2.0.6-develop.2) (2022-05-02)

## [2.0.6-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.5...v2.0.6-develop.1) (2022-05-01)

## [2.0.5](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.4...v2.0.5) (2022-04-25)

## [2.0.5-develop.3](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.5-develop.2...v2.0.5-develop.3) (2022-04-25)

## [2.0.5-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.5-develop.1...v2.0.5-develop.2) (2022-04-25)

## [2.0.5-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.4...v2.0.5-develop.1) (2022-04-06)

## [2.0.4](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.3...v2.0.4) (2022-03-31)

## [2.0.4-develop.2](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.4-develop.1...v2.0.4-develop.2) (2022-03-31)

## [2.0.4-develop.1](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.3...v2.0.4-develop.1) (2022-03-31)

## [2.0.3](https://github.com/sebbo2002/ble2mqtt/compare/v2.0.2...v2.0.3) (2022-03-01)


### Bug Fixes

* Handle null/undefined return type properly ([59b415f](https://github.com/sebbo2002/ble2mqtt/commit/59b415f7134fd574831f8906cd8b6c53751b90ec))
* **SwitchBot:** Handle longer curtain advertisements as well ([32c9174](https://github.com/sebbo2002/ble2mqtt/commit/32c91748896a9c5767c451bcf387e91e1972dd4a))


modules/mi-flora.js
'use strict';

const MiFloraServiceUUID = 'fe95';
const MiFloraSensorTypes = {
    '04': 'temperature',
    '07': 'illuminance',
    '08': 'moisture',
    '09': 'fertility'
};

module.exports = class MiFloraSensorModule {
    static getModuleDescription () {
        return {
            name: 'mi-flora',
            serviceUUIDs: [MiFloraServiceUUID]
        };
    }

    static handleAdvertisement ({debug, serviceData}) {
        const dataObj = serviceData.find(({uuid}) => uuid === MiFloraServiceUUID);
        if (!dataObj) {
            debug('Got an advertisement, but there\'s no service data attached. Ignore it.');
            return;
        }

        const data = dataObj.data;
        if (data.length < 14) {
            debug(`Got an advertisement, but service data is too short (${data.length}). Do you have the current firmware installed?`);
            return;
        }

        const type = data.toString('hex', 12, 13);
        if (!['04', '07', '08', '09'].includes(type)) {
            debug(`Got an advertisement, but service data seems to be invalid (not a known type: ${type})`);
            return;
        }

        const valueLength = parseInt(data.toString('hex', 14, 15));
        if (data.length - valueLength !== 15) {
            debug('Got an advertisement, but service data seems to be invalid (not a known length)');
            return;
        }

        const reversedBuffer = new Buffer.from(
            data.toString('hex', 15, 15 + valueLength)
                .match(/.{1,2}/g)
                .reverse()
                .join(''),
            'hex'
        );

        const value = parseInt(reversedBuffer.toString('hex'), 16);
        return {
            [MiFloraSensorTypes[type]]: (type === '04' ? (value * 0.1).toFixed(1) : value)
        };
    }
};
